Author: Andreas Beckmann <anbe@debian.org>
Description: correct the usage of _POSIX_V6_LP64_OFF64 etc.
 which caused a 32bit long being used for POV_LONG while this is required
 to be a 64bit integer
 integer overflow during sentinel conversion lead to segmentation faults

if _POSIX_V6_LP*_OFF* is defined to -1 the feature is *not* available
(as seen on most 32bit platforms: arm*, mips*, (!linux)-i386)
if it's undefined, it should be checked with e.g.
  if (__sysconf (_SC_V7_LP64_OFF64) > 0)
see e.g. eglibc source posix/confstr.c

*** Note: his patch is not complete, probably all other uses of
***      _POSIX_* macros need the same treatment

Bug-Debian: https://bugs.debian.org/745448
Bug: https://github.com/POV-Ray/povray/issues/27

--- a/vfe/unix/syspovconfig_bsd.h
+++ b/vfe/unix/syspovconfig_bsd.h
@@ -52,7 +52,8 @@
 #endif
 
 // TODO - the POV_LONG stuff is just copied from the Posix settings; someone needs to verify universal BSD compatibility.
-#if defined(_POSIX_V6_LPBIG_OFFBIG) || defined(_POSIX_V6_LP64_OFF64)
+#if (defined(_POSIX_V6_LPBIG_OFFBIG) && _POSIX_V6_LPBIG_OFFBIG > 0) \
+    || (defined(_POSIX_V6_LP64_OFF64) && _POSIX_V6_LP64_OFF64 > 0)
 	// long is at least 64 bits.
 	#define POV_LONG long
 #elif defined(_POSIX_V6_ILP32_OFFBIG) || defined(_POSIX_V6_ILP32_OFF32)
--- a/vfe/unix/syspovconfig_linux.h
+++ b/vfe/unix/syspovconfig_linux.h
@@ -39,7 +39,8 @@
 
 // lseek64 is natively supported on GNU/Linux systems.
 
-#if defined(_POSIX_V6_LPBIG_OFFBIG) || defined(_POSIX_V6_LP64_OFF64)
+#if (defined(_POSIX_V6_LPBIG_OFFBIG) && _POSIX_V6_LPBIG_OFFBIG > 0) \
+    || (defined(_POSIX_V6_LP64_OFF64) && _POSIX_V6_LP64_OFF64 > 0)
 	// long is at least 64 bits.
 	#define POV_LONG long
 #elif defined(_POSIX_V6_ILP32_OFFBIG) || defined(_POSIX_V6_ILP32_OFF32)
--- a/vfe/unix/syspovconfig_osx.h
+++ b/vfe/unix/syspovconfig_osx.h
@@ -41,7 +41,8 @@
 #define lseek64(handle,offset,whence) lseek(handle,offset,whence)
 
 // TODO - the POV_LONG stuff is just copied from the Posix settings; someone needs to test this on OS X.
-#if defined(_POSIX_V6_LPBIG_OFFBIG) || defined(_POSIX_V6_LP64_OFF64)
+#if (defined(_POSIX_V6_LPBIG_OFFBIG) && _POSIX_V6_LPBIG_OFFBIG > 0) \
+    || (defined(_POSIX_V6_LP64_OFF64) && _POSIX_V6_LP64_OFF64 > 0)
 	// long is at least 64 bits.
 	#define POV_LONG long
 #elif defined(_POSIX_V6_ILP32_OFFBIG) || defined(_POSIX_V6_ILP32_OFF32)
--- a/vfe/unix/syspovconfig_posix.h
+++ b/vfe/unix/syspovconfig_posix.h
@@ -51,7 +51,8 @@
 	#define lseek64(handle,offset,whence) lseek(handle,offset,whence)
 #endif
 
-#if defined(_POSIX_V6_LPBIG_OFFBIG) || defined(_POSIX_V6_LP64_OFF64)
+#if (defined(_POSIX_V6_LPBIG_OFFBIG) && _POSIX_V6_LPBIG_OFFBIG > 0) \
+    || (defined(_POSIX_V6_LP64_OFF64) && _POSIX_V6_LP64_OFF64 > 0)
 	// long is at least 64 bits.
 	#define POV_LONG long
 #elif defined(_POSIX_V6_ILP32_OFFBIG) || defined(_POSIX_V6_ILP32_OFF32)
